hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/client/common/FlinkRowDataReaderContext.java

```diff
@@ -0,0 +1,255 @@
+
+    List<String> fieldNames = dataSchema.getFields().stream().map(Schema.Field::name).collect(Collectors.toList());
+    List<DataType> fieldTypes = dataSchema.getFields().stream().map(
+        f -> AvroSchemaConverter.convertToDataType(f.schema())).collect(Collectors.toList());
+    int[] selectedFields = requiredSchema.getFields().stream().map(Schema.Field::name)
+        .map(fieldNames::indexOf)
+        .mapToInt(i -> i)
+        .toArray();
+
+    return fileReader.getRowDataIterator(
+        fieldNames,
+        fieldTypes,
+        selectedFields,
+        predicates,
+        filePath,
+        start,
+        length);
+  }
+
+  @Override
+  public Option<HoodieRecordMerger> getRecordMerger(RecordMergeMode mergeMode, String mergeStrategyId, String mergeImplClasses) {
+    switch (mergeMode) {
+      case EVENT_TIME_ORDERING:
+        return Option.of(new EventTimeFlinkRecordMerger());
+      case COMMIT_TIME_ORDERING:
+        return Option.of(new CommitTimeFlinkRecordMerger());
+      default:
+        Option<HoodieRecordMerger> mergerClass =
+            HoodieRecordUtils.createValidRecordMerger(EngineType.FLINK, mergeImplClasses, mergeStrategyId);
+        if (mergerClass.isEmpty()) {
+          throw new HoodieValidationException("No valid flink merger implementation set for `"
+              + RECORD_MERGE_IMPL_CLASSES_WRITE_CONFIG_KEY + "`");
+        }
+        return mergerClass;
+    }
+  }
+
+  @Override
+  public Object getValue(RowData record, Schema schema, String fieldName) {
+    Option<RowData.FieldGetter> fieldGetterOpt = HoodieRowDataUtil.getFieldGetter(schema, fieldName);
+    return fieldGetterOpt.isEmpty() ? null : fieldGetterOpt.get().getFieldOrNull(record);
+  }
+
+  @Override
+  public String getRecordKey(RowData record, Schema schema) {
+    return Objects.toString(getValue(record, schema, RECORD_KEY_METADATA_FIELD));
+  }
+
+  @Override
+  public Map<String, Object> generateMetadataForRecord(RowData record, Schema schema, Option<String> orderingFieldName) {
+    Map<String, Object> metadata = super.generateMetadataForRecord(record, schema, orderingFieldName);
+    if (orderingFieldName.isEmpty() || metadata.containsKey(INTERNAL_META_ORDERING_FIELD)) {
+      return metadata;
+    }
+    Comparable orderingValue = getOrderingValue(Option.of(record), metadata, schema, orderingFieldName);
+    metadata.put(INTERNAL_META_ORDERING_FIELD, orderingValue);
+    return metadata;
+  }
+
+  @Override
+  public HoodieRecord<RowData> constructHoodieRecord(Option<RowData> recordOption, Map<String, Object> metadataMap) {
+    HoodieKey hoodieKey = new HoodieKey(
+        (String) metadataMap.get(INTERNAL_META_RECORD_KEY),
+        (String) metadataMap.get(INTERNAL_META_PARTITION_PATH));
+    // delete record
+    if (recordOption.isEmpty()) {
+      return new HoodieEmptyRecord<>(hoodieKey, HoodieRecord.HoodieRecordType.FLINK);
```

Comment:
do we need to set up the ordering value correctly.

Reply:
Nice catch, will update.